src/mscorlib/shared/System/UnitySerializationHolder.cs

```diff
@@ -0,0 +1,56 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.Serialization;
+using System.Reflection;
+
+namespace System
+{
+    /// <summary>
+    /// Holds Null class for which we guarantee that there is only ever one instance of.
+    /// This only exists for backwarts compatibility with 
+    /// </summary>
+#if CORECLR
+    internal
+#else
+    public  // On CoreRT this must be public.
+#endif
+    sealed class UnitySerializationHolder : ISerializable, IObjectReference
+    {
+        internal const int NullUnity = 0x0002;
+
+        public static void GetUnitySerializationInfo(SerializationInfo info, int unityType, string data, Assembly assembly)
+        {
+            // A helper method that returns the SerializationInfo that a class utilizing 
+            // UnitySerializationHelper should return from a call to GetObjectData.  It contains
+            // the unityType (defined above) and any optional data (used only for the reflection
+            // types.)
+
+            info.SetType(typeof(UnitySerializationHolder));
+            info.AddValue("Data", data, typeof(string));
+            info.AddValue("UnityType", unityType);
+            info.AddValue("AssemblyName", assembly?.FullName ?? string.Empty);
+        }
+
+        public UnitySerializationHolder(SerializationInfo info, StreamingContext context)
+        {
+            if (info == null)
+            {
+                throw new ArgumentNullException(nameof(info));
+            }
+
+            // We are ignoring any serialization input as we are only concerned about DBNull.
+        }
+
+        public void GetObjectData(SerializationInfo info, StreamingContext context) =>
+            throw new NotSupportedException(SR.NotSupported_UnitySerHolder);
+
+        public object GetRealObject(StreamingContext context)
+        {
+            // We are always returning the same DBNull instance and ignoring serialization input.
+
```

Comment:
If we're still trying to be compatible with legacy blobs, should we at least throw if "UnityType" isn't `NullUnity`?

Reply:
We definitely should throw as otherwise users would just get a random object. But before I can do that I need to verify the weird behavior I posted above. Thanks!

Reply:
The SerializationInfo isn't an ordered list - it's a dictionary of name-value pairs. So there's no problem.